lib/cloud_profiler_agent/looper.rb

```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+
+require 'google/apis/errors'
+
+module CloudProfilerAgent
+  # Looper is responsible for the main loop of the agent. It calls a
+  # block repeatedly, handling errors, backing off, and retrying as
+  # appropriate.
+  class Looper
+    def initialize(
+      min_iteration_time: 10,
+      max_iteration_time: 60 * 60,
+      backoff_factor: 1.5,
+      debug_logging: false,
+      sleeper: ->(sec) { sleep(sec) },
+      clock: -> { Process.clock_gettime(Process::CLOCK_MONOTONIC) },
+      rander: -> { rand }
+    )
+
+      # the minimum and maximum time between iterations of the profiler loop,
+      # in seconds. Normally the Cloud Profiler API tells us how fast to go,
+      # but we back off in case of error.
+      @min_iteration_time = min_iteration_time
+      @max_iteration_time = max_iteration_time
+      @backoff_factor = backoff_factor
+
+      # stubbable for testing
+      @sleeper = sleeper
+      @clock = clock
+      @rander = rander
+
+      @debug_logging = debug_logging
+    end
+
+    attr_reader :min_iteration_time, :max_iteration_time, :backoff_factor
+
+    def run(max_iterations = 0)
+      iterations = 0
+      iteration_time = @min_iteration_time
+      loop do
+        start_time = @clock.call
+        iterations += 1
+        begin
+          yield
+        rescue ::Google::Apis::ClientError => e
+          backoff = backoff_duration(e)
+          if backoff.nil?
+            iteration_time = @max_iteration_time
+          else
+            debug_log("sleeping for #{backoff} at request of server")
+            @sleeper.call(backoff)
```

Comment:
So max iteration time can be exceeded if Google asks us to back off

Reply:
yes, that's intentional. Google throttles clients to get 1 profile per minute, so with a very large number of clients they might ask us to back off for a pretty long time.